### PR TITLE
feat: implement --tee option for output logging (#352)

### LINKED ISSUE #352: Implement --tee option for output logging
## Overview

Implement the `--tee` option to append output to a file while displaying on console, compatible with Google Cloud Spanner CLI.

## Specification

- Flag: `--tee=<filename>`
- Appends a copy of all output to the specified file
- Output still appears on console (similar to Unix tee command)

## Implementation notes

- File should be opened in append mode
- Handle file creation if it doesn't exist
- Related to `\T` meta-command functionality (issue #345)
- Consider implementing shared infrastructure for both features

## Priority

Medium priority - useful for logging and audit purposes

## References

- Parent issue: #342
- Related to `\T` meta-command (issue #345)
- [Google Cloud Spanner CLI documentation](https://cloud.google.com/spanner/docs/spanner-cli)

### PR DESCRIPTION
**Pull request by apstndb**
## Summary

Implements the `--tee` option to append output to a file while displaying on console, compatible with Google Cloud Spanner CLI.

Fixes #352

## Changes

- Added `--tee` flag to CLI options that accepts a filename
- Implemented output tee functionality using `io.MultiWriter`
- Introduced stream separation architecture to handle TTY-specific operations
- Added comprehensive tests for tee functionality
- Updated README.md with --tee documentation

### Design Decisions

1. **Stream Separation Architecture**: Introduced `TtyOutStream` to separate TTY-specific operations from regular output. This ensures that terminal control sequences (progress marks, prompts) don't pollute tee files.

2. **io.MultiWriter Usage**: Used Go's standard `io.MultiWriter` instead of implementing a custom TeeWriter, keeping the implementation simple and leveraging well-tested standard library functionality.

3. **Append Mode**: File is opened with `O_APPEND` flag for consistency with typical log file behavior and to preserve existing content.

### What Gets Captured in Tee

**Included:**
- Query results and tables
- Error messages and warnings
- Result metadata (row counts, execution times)
- SQL statements when `CLI_ECHO_INPUT` is enabled

**Excluded:**
- Interactive prompts (e.g., `spanner>`)
- Progress indicators (DDL progress bars)
- Confirmation dialogs (DROP DATABASE confirmations)
- Readline input display
- Any output using carriage returns (`\r`)

### Key Implementation Points

1. **Terminal Size Detection**: Modified to use `TtyOutStream` when available via `GetTerminalSizeWithTty()`, ensuring AUTOWRAP continues to work correctly even when output is an `io.MultiWriter`.

2. **Readline Integration**: Configured readline's `LineEditor.Writer` to use `TtyOutStream` to prevent prompts from appearing in tee files.

3. **Progress Mark Handling**: All progress indicators now route through `TtyOutStream` to avoid carriage return pollution in logs.

4. **CLI_ECHO_INPUT Behavior**: Echoed SQL statements are intentionally included in tee output to provide complete context in logs.

## Testing

- Added unit tests to verify stream separation behavior
- Tests confirm progress marks don't appear in tee files
- Terminal size detection tests with MultiWriter
- Non-TTY environment handling for CI compatibility

## Example Usage

```bash
# Log all query results to a file
$ spanner-mycli --tee output.log -p myproject -i myinstance -d mydb

# In batch mode with --tee
$ spanner-mycli --tee queries.log -p myproject -i myinstance -d mydb -e 'SELECT * FROM users;'

# With CLI_ECHO_INPUT enabled
spanner> SET CLI_ECHO_INPUT = TRUE;
spanner> SELECT 1 AS test;
# Both the query and results appear in the tee file
```
EOF < /dev/null

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -473,7 +474,30 @@ func run(ctx context.Context, opts *spannerOptions) error {
...
+		// Open tee file in append mode (creates if doesn't exist)
+		teeFile, err := os.OpenFile(opts.Tee, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
+		if err != nil {
+			return fmt.Errorf("failed to open tee file %q: %w", opts.Tee, err)
+		}
+		defer teeFile.Close()
+		
+		// Create a MultiWriter that writes to both stdout and the tee file
+		// All normal output (query results, messages) will be captured in the tee file
+		outStream = io.MultiWriter(os.Stdout, teeFile)
```
[the remark this answers is not in the document]
Comment (apstndb) [reply main.go]: Implemented safeTeeWriter that handles write errors gracefully. On write failure, it prints a warning once to stderr and then discards subsequent writes to prevent interrupting stdout. Added comprehensive unit tests.

Fixed in commit 219c99f.

File: integration_tee_test.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+package main
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+)
+
+func TestTeeOptionIntegration(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (apstndb) [reply integration_tee_test.go]: Removed the incomplete integration test file. Will implement proper integration tests in a future PR when the test infrastructure is ready.

Fixed in commit 219c99f.

**Comment by apstndb** [comment]
All review feedback has been addressed:

✅ Implemented safeTeeWriter for graceful error handling
✅ Removed incomplete integration tests
✅ Added validation to prevent hanging on non-regular files (pipes, directories)

All tests are passing and the PR is ready to merge.

File: system_variables.go [hunk 3]
```diff
@@ -1072,7 +1087,12 @@ var systemVariableDefMap = map[string]systemVariableDef{
 		Description: "Get the current screen width in spanner-mycli client-side statement.",
 		Accessor: accessor{
 			Getter: func(this *systemVariables, name string) (map[string]string, error) {
-				width, err := GetTerminalSize(this.CurrentOutStream)
+				// Prefer TtyOutStream for terminal size detection
+				w := this.CurrentOutStream
+				if this.TtyOutStream != nil {
+					w = this.TtyOutStream
+				}
+				width, err := GetTerminalSize(w)
```
[the remark this answers is not in the document]
Comment (apstndb) [reply system_variables.go]: Fixed the terminal width detection to handle MultiWriter case. When --tee is enabled, CurrentOutStream is an io.MultiWriter which cannot be cast to *os.File. The getter now uses TtyOutStream when available, falling back to type assertion only when TtyOutStream is nil. Added tests to verify the fix.

Fixed in commit 76ab52b.

File: main.go [hunk 4]
```diff
@@ -474,7 +509,49 @@ func run(ctx context.Context, opts *spannerOptions) error {
 		sysVars.WithoutAuthentication = true
 	}
 
-	cli, err := NewCli(ctx, cred, os.Stdin, os.Stdout, os.Stderr, &sysVars)
+	// Always keep the original os.Stdout for TTY operations
+	// This is used for progress marks, readline prompts, and other terminal-specific output
+	// that should not be captured in the tee file
+	sysVars.TtyOutStream = os.Stdout
```
[the remark this answers is not in the document]
Comment (apstndb) [reply main.go]: Fixed to only set TtyOutStream when stdout is actually a terminal using term.IsTerminal(). This prevents writing terminal control codes to redirected files and ensures proper behavior when stdout is not a TTY.

Fixed in commit a355167.

File: cli.go [hunk 5]
```diff
@@ -208,7 +222,12 @@ func (c *Cli) handleSpecialStatements(ctx context.Context, stmt Statement) (exit
 			return -1, true
 		}
 
-		if !confirm(c.InStream, c.OutStream, fmt.Sprintf("Database %q will be dropped.\nDo you want to continue?", s.DatabaseId)) {
+		// Use TtyOutStream for interactive confirmations to avoid polluting tee output
+		confirmOut := c.OutStream
+		if c.SystemVariables.TtyOutStream != nil {
+			confirmOut = c.SystemVariables.TtyOutStream
+		}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply cli.go]: Fixed to check if TtyOutStream is nil and show an appropriate error message when trying to confirm DROP DATABASE without a TTY. This prevents hanging when stdout is redirected.

Fixed in commit a355167.

File: cli_readline.go [hunk 6]
```diff
@@ -114,7 +114,12 @@ func initializeMultilineEditor(c *Cli) (*multiline.Editor, History, error) {
 	ed := &multiline.Editor{}
 	
 	// Configure the LineEditor with proper I/O streams
-	ed.LineEditor.Writer = c.OutStream
+	// Use TtyOutStream for readline to avoid polluting tee output with prompts
+	if c.SystemVariables.TtyOutStream != nil {
+		ed.LineEditor.Writer = c.SystemVariables.TtyOutStream
+	} else {
+		ed.LineEditor.Writer = c.OutStream
+	}
```
[the remark this answers is not in the document]
Comment (apstndb) [reply cli_readline.go]: Fixed to return an error when trying to run in interactive mode without a TTY. The error message clearly indicates that stdout is not a terminal, preventing readline from writing prompts to redirected files.

Fixed in commit a355167.

**Comment by apstndb** [comment]
## Summary of Completed Work

The `--tee` option implementation for issue #352 is now complete and ready for merge\! 

### What Was Implemented

1. **Core Functionality**:
   - Added `--tee <filename>` flag that appends output to a file while displaying on console
   - Uses `io.MultiWriter` for efficient simultaneous output
   - Creates the file if it doesn't exist, appends if it does
   - Compatible with Google Cloud Spanner CLI behavior

2. **Stream Separation Architecture**:
   - Introduced `TtyOutStream` for TTY-specific operations (prompts, progress marks)
   - Keeps tee files clean by excluding terminal control sequences
   - Properly handles `CLI_ECHO_INPUT` to include echoed queries in tee logs

3. **Robust Error Handling**:
   - Implemented `safeTeeWriter` that handles write failures gracefully
   - Prevents hanging on special files (FIFOs, device files)
   - Properly handles redirected stdout scenarios

4. **Terminal Width Detection**:
   - Fixed `CLI_CURRENT_WIDTH` to work with MultiWriter
   - Ensures AUTOWRAP functionality works correctly with --tee

5. **Non-TTY Environment Support**:
   - Only sets TtyOutStream when stdout is actually a terminal
   - Disables progress marks when no TTY available
   - Prevents interactive mode from starting without TTY
   - Shows appropriate errors for operations requiring TTY

### Testing
- Added comprehensive unit tests covering all scenarios
- Tests pass in both TTY and non-TTY environments
- All existing tests continue to pass

### Documentation
- Updated README.md with --tee usage and behavior
- Added architecture documentation explaining stream separation
- Documented all edge cases and design decisions

### Review Feedback Addressed
All review feedback from Gemini Code Assist has been addressed:
- ✅ Error handling for tee file writes
- ✅ Protection against hanging on special files
- ✅ Robust terminal width detection
- ✅ Proper handling of redirected stdout

The PR is clean, all checks pass, and it's ready to merge\!

File: main.go [hunk 7]
```diff
@@ -474,7 +509,50 @@ func run(ctx context.Context, opts *spannerOptions) error {
...
+		// Check if the file exists and is a regular file before opening.
+		// This prevents blocking on special files like FIFOs.
+		fi, err := os.Stat(opts.Tee)
+		if err != nil && !os.IsNotExist(err) {
+			// An error other than "not found" occurred.
+			return fmt.Errorf("failed to stat tee file %q: %w", opts.Tee, err)
+		}
+		if err == nil && !fi.Mode().IsRegular() {
+			// The file exists but is not a regular file.
+			return fmt.Errorf("tee output to a non-regular file is not supported: %s", opts.Tee)
```
[the remark this answers is not in the document]
Comment (apstndb) [reply main.go]: Thank you for the feedback. While the original logic was actually correct (the condition IS reachable when file exists), I agree it was confusing. 

I've refactored the code to use a switch statement that makes the three cases explicit:
1. File exists → check if regular
2. File doesn't exist → will be created
3. Other stat error → return error

The new implementation is clearer and avoids the double-negative condition that caused confusion.

Fixed in commit 022d944.
